# This file contains pipeline constants and a few functions.

SAMPLE_ID = config["input"]["id"]
WORKDIR = os.path.join(config["workdir"], SAMPLE_ID)
SUPPORTED_FILETYPES = {".fastq.gz", ".fastq", ".bam"}

# TODO(julia): this is specific to GRCh37, populate these from the reference genome instead
_CHR = [str(x) for x in range(1, 23)] + ['X', 'Y', 'MT']

# will default to false, if not present in config
_PARALLEL_INDEL_REALIGNER = config.get("parallel_indel_realigner")

_READ_LENGTH = 124
_STAR_GENOME_DIR = os.path.join(
    os.path.dirname(config["reference"]["genome"]), "star-genome-%d" % _READ_LENGTH)

# Common functions

def _get_half_cores(_):
  return max(1, int(config["num_threads"]/2))

def _get_all_cores(_):
  return int(config["num_threads"])

def _get_intervals_str(wildcards):
  return "--intervals %s" % wildcards.chr if wildcards.chr in _CHR else ""

rule gunzip:
  input:
    "{prefix}.{ext}.gz"
  output:
    "{prefix}.{ext}"
  wildcard_constraints:
    ext="(gtf|vcf|fasta)"
  shell:
    "gunzip -c {input} > {output}"
